tratto/src/main/java/star/tratto/util/StringUtils.java

```diff
@@ -89,4 +112,90 @@ public static String fullyQualifiedClassName(Pair<String, String> packageClassPa
         }
         return packageClassPair.getValue0() + "." + packageClassPair.getValue1();
     }
+
+    /**
+     * Computes the semantic similarity of two strings by the cosine
+     * similarity of word frequencies in the input.
+     * NOTE: During semantic comparison, a word such as "text2int"
+     * will be considered as a single world "textint" rather than two words,
+     * "text" and "int". This avoids potential oversimplification.
+     *
+     * @param s1 a string of one or more words separated by a space
+     * @param s2 a string of one or more words separated by a space
+     * @return the cosine similarity of the two strings represented by a
+     * double between 0.0 and 1.0
+     */
+    public static double semanticSimilarity(String s1, String s2) {
+        s1 = s1.replaceAll("[^a-zA-Z ]", "").toLowerCase();
+        s2 = s2.replaceAll("[^a-zA-Z ]", "").toLowerCase();
+        // Convert the words into lemmas.
+        List<String> tokens1 = lemmatize(stanfordCoreNLP, s1);
+        List<String> tokens2 = lemmatize(stanfordCoreNLP, s2);
+        // Compute the cosine similarity of the two vectors of word frequencies.
+        return cosineSimilarity(tokens1, tokens2);
+    }
+
+    /**
+     * Converts a string into a list of lemmas using the CoreNLP library.
+     */
+    private static List<String> lemmatize(StanfordCoreNLP pipeline, String documentText) {
+        Annotation document = new Annotation(documentText);
+        pipeline.annotate(document);
+        List<CoreLabel> tokens = document.get(TokensAnnotation.class);
+        return tokens
+                .stream()
+                .map(t -> t.get(LemmaAnnotation.class))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Computes the cosine similarity of two lists of words.
+     *
+     * @param list1 list of words
+     * @param list2 list of words
+     * @return the cosine similarity (double between 0.0 and 1.0)
+     */
+    private static double cosineSimilarity(List<String> list1, List<String> list2) {
+        Map<String, Integer> map1 = wordFrequencies(list1);
+        Map<String, Integer> map2 = wordFrequencies(list2);
+        LinkedHashSet<String> intersection = new LinkedHashSet<>(map1.keySet());
```

Comment:
Consider using TreeSet, or sorting map.keySet(), to make your code deterministic.